### PR TITLE
feat: add keccak builtin

### LINKED ISSUE #69: feat: implement Keccak builtin

### PR DESCRIPTION
**Pull request by zmalatrax**
Closes #69

### CODE REVIEW

File: src/builtins/keccak.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      ...target.slice(inputOffset, outputOffset).map((value) => {
+        if (!isFelt(value)) throw new ExpectedFelt();
+        if (value.toBigInt() >> KECCAK_BITS !== 0n) throw new Error();
+        return numberToBytesLE(value.toBigInt(), 25);
+      })
+    );
+
+    const state = u32(input);
+    keccakP(state);
+    const finalState = u8(state);
```
Comment (ClementWalter) [inline src/builtins/keccak.ts:49]: is `keccakP` modifying the `const` state?

File: src/builtins/keccak.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    const outputs = Array.from({ length: 8 }, (_, i) => {
+      const size = i * KECCAK_BYTES;
+      return finalState.slice(size, size + KECCAK_BYTES);
+    }).map(bytesToNumberLE);
```
Comment (ClementWalter) [inline src/builtins/keccak.ts:54]: I'd go for somethhing more like

```
(new Array(8)).map((i) => finalState.slice(i * k, (i + 1) * k)).map(bytesToNumberLE)
```

the fact that we just split the finalState in chunks is not visible here

File: src/builtins/keccak.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      case 5:
+        target[offset] = new Felt(outputs[5]);
+        break;
+      case 6:
+        target[offset] = new Felt(outputs[6]);
+        break;
+      default:
+        target[offset] = new Felt(outputs[7]);
+        break;
+    }
```
Comment (ClementWalter) [inline src/builtins/keccak.ts:81]: why not `target[offset] = new Felt(outputs[keccakIndex - inputCellsPerKeccak]);`

File: src/builtins/keccak.ts (hunk 1 above)
Comment (zmalatrax) [reply src/builtins/keccak.ts:49]: Yes, `keccakP` modifies `state`. The `const` keyword blocks re-declaration of objects (i.e. cannot do `state = newState`) but object properties can still be mutated.

Even if `state` values are mutated, `const` keyword avoids assignment overwriting, but `let` could be used for clarity (gotta be consistent throughout the codebase)

Quoting [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const)
> The const declaration creates an immutable reference to a value. It does not mean the value it holds is immutable — just that the variable identifier cannot be reassigned. For instance, in the case where the content is an object, this means the object's contents (e.g., its properties) can be altered. You should understand const declarations as "create a variable whose identity remains constant", not "whose value remains constant" — or, "create immutable [bindings](https://developer.mozilla.org/en-US/docs/Glossary/Binding)", not "immutable values".
>
> Many style guides (including [MDN's](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Writing_style_guide/Code_style_guide/JavaScript#variable_declarations)) recommend using const over [let](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let) whenever a variable is not reassigned in its scope. This makes the intent clear that a variable's type (or value, in the case of a primitive) can never change. Others may prefer let for non-primitives that are mutated.

File: src/builtins/keccak.ts (hunk 2 above)
Comment (zmalatrax) [reply src/builtins/keccak.ts:54]: Creating a new array with `new Array(8)` will create an array of length 8 filled with empty items, which means that `.map` will not iterate over them (thus outputs still being an empty array after the `map`)

**Review by ClementWalter — DISMISSED** [review]

File: src/builtins/keccak.ts (hunk 2 above)
Comment (ClementWalter) [reply src/builtins/keccak.ts]: indeed, should be `(new Array(8)).fill(0)` which is not better then

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**